### PR TITLE
Add a new "system call" in a namespace parallel to WASI.

### PR DESCRIPTION
**Pull request by egrimley-arm**
The new function, called fd_create, in a separate module, called veracruz_si, creates an anonymous file. There are test programs in C and Rust, and the Rust test program is included in the CI tests.

### CODE REVIEW

**Review by ShaleXIONG — APPROVED** [review]

File: sdk/fd_create/prog.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
+// This test program is not ready to be released!
```
Comment (dominic-mulligan-arm) [inline sdk/fd_create/prog.c:1]: I think this is the first small C example we have.  Can you put it under `c-examples`, in analogy to `rust-examples` that already exists in the SDK?

Comment (dominic-mulligan-arm) [inline sdk/fd_create/prog.c:1]: Also: can you add our standard copyright/licensing header to the top of this file?

File: sdk/fd_create/test.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+# This test script is not ready to be released!
```
Comment (dominic-mulligan-arm) [inline sdk/fd_create/test.sh:3]: Can you add the copyright/licensing header to this.  Also, why is this file needed rather than e.g. hooking the example into our integration tests?

File: sdk/rust-examples/fd_create/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "fd_create"
+version = "0.1.0"
+authors = ["Edmund Grimley Evans <edmund.grimley-evans@arm.com>"]
```
Comment (dominic-mulligan-arm) [inline sdk/rust-examples/fd_create/Cargo.toml:4]: Can you attribute to the "Veracruz development team" and set the version as "0.3.0".  Also, can you provide a "description" field?

File: sdk/rust-examples/fd_create/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+// This test program is not ready to be released!
```
Comment (dominic-mulligan-arm) [inline sdk/rust-examples/fd_create/src/main.rs:1]: Can you add the licensing/copyright header?

File: sdk/rust-examples/fd_create/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+#![feature(wasi_ext)]
+
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::SeekFrom;
+use std::os::wasi::io::{FromRawFd, RawFd};
+
+////////////////////////////////////////////////////////////////////////////
+// This part should probably be in a Veracruz library.
```
Comment (dominic-mulligan-arm) [inline sdk/rust-examples/fd_create/src/main.rs:11]: Can you move these into a separate support library that's linked against?

File: sdk/rust-examples/fd_create/test.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+# This test script is not ready to be released!
```
Comment (dominic-mulligan-arm) [inline sdk/rust-examples/fd_create/test.sh:3]: Why is this file needed?

File: sdk/rust-examples/fd_create/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
+// This test program is not ready to be released!
+
+#![feature(wasi_ext)]
+
+use std::fs::File;
+use std::io::prelude::*;
```
Comment (dominic-mulligan-arm) [inline sdk/rust-examples/fd_create/src/main.rs:6]: Has `rustfmt` been run on this file?

File: execution-engine/src/wasi/wasmi.rs [hunk 8]
```diff
@@ -26,6 +26,47 @@ use wasmi::{
...
+
+impl ToPrimitive for APIName {
+    fn to_i64(&self) -> Option<i64> {
+        Some(self.to_u64().unwrap() as i64)
+    }
+    fn to_u64(&self) -> Option<u64> {
+        match self {
+            APIName::WasiAPIName(WasiAPIName::_LAST) => unreachable!(),
+            APIName::WasiAPIName(x) => x.to_u64(),
+            APIName::VeracruzAPIName(x) => Some(WasiAPIName::_LAST as u64 + x.to_u64().unwrap()),
```
Comment (dominic-mulligan-arm) [inline execution-engine/src/wasi/wasmi.rs:65]: Can this `unwrap()` abort at runtime?

File: execution-engine/src/wasi/wasmi.rs [hunk 9]
```diff
@@ -386,20 +437,27 @@ impl ModuleImportResolver for WASMIRuntimeState {
...
         if !TypeCheck::check_signature(index, signature) {
             Err(Error::Instantiation(format!(
                 "Function {} has an unexpected type-signature: {:?}.",
                 field_name, signature
             )))
         } else {
-            Ok(FuncInstance::alloc_host(signature.clone(), index as usize))
+            Ok(FuncInstance::alloc_host(
+                signature.clone(),
+                index.to_usize().unwrap(),
```
Comment (dominic-mulligan-arm) [inline execution-engine/src/wasi/wasmi.rs:459]: Can this `unwrap()` abort at runtime?

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]

**Comment by egrimley-arm** [comment]
I've separated the plumbing from the output and put the plumbing in https://github.com/veracruz-project/veracruz/pull/308.

**Comment by dominic-mulligan-arm** [comment]
What is the status of this PR, @egrimley-arm?

**Comment by egrimley-arm** [comment]
> What is the status of this PR, @egrimley-arm?

I think I'm going to improve the tests and then mark it "ready for review".

File: sdk/rust-examples/fd_create/test.sh (hunk 6 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/test.sh]: It's not really needed any more as I've included the program in the standard tests.

File: sdk/rust-examples/fd_create/src/main.rs (hunk 5 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/src/main.rs:23]: I don't know. I could investigate at some point.

File: sdk/rust-examples/fd_create/src/main.rs (hunk 4 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/src/main.rs]: Done.

File: sdk/fd_create/prog.c (hunk 1 above)
Comment (egrimley-arm) [reply sdk/fd_create/prog.c]: Done.

Comment (egrimley-arm) [reply sdk/fd_create/prog.c]: Done.

File: sdk/fd_create/test.sh (hunk 2 above)
Comment (egrimley-arm) [reply sdk/fd_create/test.sh]: We don't have a C toolchain available and I'm not sure if we want to add that dependency yet.

File: sdk/rust-examples/fd_create/Cargo.toml (hunk 3 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/Cargo.toml:4]: Done.

File: sdk/rust-examples/fd_create/test.sh (hunk 6 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/test.sh]: I could delete it?

File: execution-engine/src/wasi/wasmi.rs (hunk 9 above)
Comment (egrimley-arm) [reply execution-engine/src/wasi/wasmi.rs:486]: I don't think so. Is it worth investigating?

File: sdk/rust-examples/fd_create/src/main.rs (hunk 7 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/src/main.rs:18]: Done.

File: execution-engine/src/wasi/wasmi.rs (hunk 9 above)
Comment (egrimley-arm) [reply execution-engine/src/wasi/wasmi.rs:486]: The default conversion functions are used for `WasiAPIName` and `VeracruzAPIName`, so presumably they return a non-negative number that can't be too big for a `usize`.

File: sdk/rust-examples/fd_create/src/main.rs (hunk 5 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/src/main.rs:23]: It really ought to be possible: I don't see any problem, apart from the usual one of deciding what the crate should be called and where it should live in the file tree. It could be called `veracruz_app_lib`, for example, and live in `sdk/veracruz_app_lib`?

Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/src/main.rs]: That's done now.

File: sdk/rust-examples/fd_create/rust-toolchain [hunk 10]
```diff
@@ -0,0 +1 @@
+nightly-2021-11-01
```
Comment (ShaleXIONG) [inline sdk/rust-examples/fd_create/rust-toolchain:1]: Is it necessary here to force a potentially different version of rust?

File: sdk/c-examples/README.txt [hunk 11]
```diff
@@ -0,0 +1,7 @@
+Currently no C applications are built automatically as we have chosen
```
Comment (dominic-mulligan-arm) [inline sdk/c-examples/README.txt:1]: Can this be converted to a Markdown file?  We've no other `.txt` files in the repository.

File: sdk/veracruz-app-lib/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+[package]
```
Comment (dominic-mulligan-arm) [inline sdk/veracruz-app-lib/Cargo.toml:1]: Can we rename `veracruz-app-lib` to `libveracruz`?  This was the original name of the programming support library in Veracruz that was eliminated just before you joined the project, I think.  Moreover, as this is the Rust support library, can this be put under a Rust-specific directory in the `sdk` directory, as C/C++ will also need some analogue of this library at some point?

File: sdk/veracruz-app-lib/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    }
+}
+
+mod veracruz_si {
+    pub fn fd_create(fd: *mut crate::RawFd) -> u32 {
+        unsafe { crate::veracruz_si_import::fd_create(fd as u32) }
+    }
+}
+
+pub fn fd_create() -> std::io::Result<File> {
```
Comment (dominic-mulligan-arm) [inline sdk/veracruz-app-lib/src/lib.rs:27]: Can you add a `rustdoc` comment describing what `fd_create` does and its failure modes?

**Review by dominic-mulligan-arm — CHANGES_REQUESTED** [review]

**Comment by dominic-mulligan-arm** [comment]
Looks good to me, barring some small issues.

File: sdk/rust-examples/fd_create/rust-toolchain (hunk 10 above)
Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/rust-toolchain:1]: It's the nightly equivalent. https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1561-2021-11-01 :
"Version 1.56.1 (2021-11-01)"

We need nightly for the `wasm_import_module` stuff.

Comment (egrimley-arm) [reply sdk/rust-examples/fd_create/rust-toolchain:1]: Though bizarrely I can't now get an error if I remove that file so I'll try without it ...

File: sdk/c-examples/README.txt (hunk 11 above)
Comment (egrimley-arm) [reply sdk/c-examples/README.txt:1]: Done.

File: sdk/veracruz-app-lib/src/lib.rs (hunk 13 above)
Comment (egrimley-arm) [reply sdk/veracruz-app-lib/src/lib.rs:27]: Done.

File: sdk/veracruz-app-lib/Cargo.toml (hunk 12 above)
Comment (egrimley-arm) [reply sdk/veracruz-app-lib/Cargo.toml]: Done.

**Review by dominic-mulligan-arm — APPROVED** [review]
